scipio/src/controllers/deadline_queue.rs

```diff
@@ -0,0 +1,470 @@
+use crate::{enclose, task};
+use crate::{Latency, Local, Shares, SharesManager, TaskQueueHandle};
+use core::fmt::Debug;
+use futures::stream::StreamExt;
+use std::cell::{Cell, RefCell};
+use std::collections::VecDeque;
+use std::future::Future;
+use std::io;
+use std::pin::Pin;
+use std::rc::Rc;
+use std::time::{Duration, Instant};
+
+/// Items going into the [`DeadlineQueue`] must implement this trait.
+///
+/// It allows the [`DeadlineQueue`] to understand the progress and expectations
+/// of processing this item
+///
+/// [`DeadlineQueue`]: struct.DeadlineQueue.html
+pub trait DeadlineSource {
+    /// What type is returned by the [`action`] method
+    ///
+    /// [`action`]: trait.DeadlineSource.html#method.action
+    type Output;
+
+    /// Returns a [`Duration`] indicating when we would like this operation to complete.
+    ///
+    /// It is calculated from the point of Queueing, not from the point in which the operation
+    /// starts.
+    fn deadline(&self) -> Duration;
+
+    /// The action to execute. Usually your struct will implement an async function that you want
+    /// to see completed at a particular deadline, and then the implementation of this would be:
+    ///
+    /// ```ignore
+    /// fn action(&self) -> Pin<Box<dyn Future<Output = io::Result<Duration>> + '_>> {
+    ///    Box::pin(self.my_action())
+    /// }
+    /// ```
+    fn action(&self) -> Pin<Box<dyn Future<Output = io::Result<Self::Output>> + '_>>;
+
+    /// The total amount of units to be processed.
+    ///
+    /// This could be anything you want:
+    /// * If you are flushing a file, this could indicate the size in bytes of the buffer
+    /// * If you are scanning an array, this could be the number of elements.
+    ///
+    /// As long as this quantity is consistent with [`processed_units`] the controllers should work.
+    ///
+    /// This need not be static. On the contrary: as you are filling a new buffer you can already
+    /// add it to the queue and increase its total units as the buffer is written to. This can
+    /// lead to smoother operation as opposed to just adding a lot of units at once.
+    ///
+    /// [`processed_units`]: trait.DeadlineSource.html#method.processed_units
+    fn total_units(&self) -> u64;
+
+    /// The amount of units that were already processed.
+    ///
+    /// The units should match the quantities specified in [`total_units`].
+    /// The more often the system is made aware of processed units, the smoother the
+    /// controller will be.
+    ///
+    /// For example, you can buffer all updates and just inform that processed_units == total_units
+    /// at the end of the process, but then the controller would be a step function.
+    ///
+    /// [`total_units`]: trait.DeadlineSource.html#method.total_units
+    fn processed_units(&self) -> u64;
+}
+
+impl<T> Debug for dyn DeadlineSource<Output = T> {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
```

Comment:
Minor style nit, but I find the following way to write this to be aesthetically nicest:
```suggestion
impl<T> fmt::Debug for dyn DeadlineSource<Output = T> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```

Reply:
agree